source/shared/cpp/ObjectModel/TextBlockText.cpp

```diff
@@ -0,0 +1,243 @@
+    std::smatch matches;
+    std::string text = in;
+    enum MatchIndex
+    {
+        IsDate = 2,
+        Year = 4,
+        Month,
+        Day,
+        Hour,
+        Min,
+        Sec,
+        TimeZone = 12,
+        TZHr,
+        TZMn,
+        Format,
+        Style,
+    };
+    std::vector<int> indexer = {Year, Month, Day, Hour, Min, Sec, TZHr, TZMn};
+
+    while (std::regex_search(text, matches, pattern))
+    {
+        time_t offset{};
+        int  formatStyle{};
+        // Date is matched
+        bool isDate = matches[IsDate].matched;
+        int hours{}, minutes{};
+        struct tm parsedTm{};
+        int *addrs[] = {&parsedTm.tm_year, &parsedTm.tm_mon,
+            &parsedTm.tm_mday, &parsedTm.tm_hour, &parsedTm.tm_min,
+            &parsedTm.tm_sec, &hours, &minutes};
+
+        if (matches[Style].matched)
+        {
+            // match for long/short/compact
+            bool formatHasSpace = matches[Format].str()[1] == ' ';
+            int formatStartIndex = formatHasSpace ? 2 : 1;
+            formatStyle = matches[Format].str()[formatStartIndex];
+        }
+
+        AddTextSection(matches.prefix().str(), TextSectionFormat::RegularString);
+
+        if (!isDate && formatStyle)
+        {
+            AddTextSection(matches[0].str(), TextSectionFormat::RegularString);
+            text = matches.suffix().str();
+            continue;
+        }
+
+        for (unsigned int idx = 0; idx < indexer.size(); idx++)
+        {
+            if (matches[indexer[idx]].matched)
+            {
+                // get indexes for time attributes to index into conrresponding matches
+                // and covert it to string
+                *addrs[idx] = stoi(matches[indexer[idx]]);
+            }
+        }
+
+        // check for date and time validation
+        if (IsValidTimeAndDate(parsedTm, hours, minutes))
+        {
+            // maches offset sign, 
+            // Z == UTC, 
+            // + == time added from UTC
+            // - == time subtracted from UTC
+            if (matches[TimeZone].matched)
+            {
+                // converts to seconds
+                hours *= 3600;
+                minutes *= 60;
+                offset = hours + minutes;
+
+                wchar_t zone = matches[TimeZone].str()[0];
+                // time zone offset calculation 
+                if (zone == '+')
+                {
+                    offset *= -1;
+                }
+            }
+
+            // measured from year 1900
+            parsedTm.tm_year -= 1900;
+            parsedTm.tm_mon -= 1;
+
+            time_t utc{};
+            // converts to ticks in UTC
+            utc = mktime(&parsedTm);
+            if (utc == -1)
+            {
+                AddTextSection(matches[0], TextSectionFormat::RegularString);
```

Comment:
Is this needed to be saved?

Reply:
This is saved if the date complies with a correct formatting (regex) but the make time function fails to generate the date as expected, so the original string is saved to be printed

Reply:
i think localtime function will fail? if utc == -1, and it can be caught in the else statement?

---
In reply to: [168269856](https://github.com/Microsoft/AdaptiveCards/pull/1102#discussion_r168269856) [](ancestors = 168269856)